flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableInputFormat.java

```diff
@@ -207,19 +211,22 @@ public Row nextRecord(Row ignore) throws IOException {
 		if (reachedEnd()) {
 			return null;
 		}
-		Row row = new Row(rowArity);
+		Row row = new Row(fields.length);
 		try {
 			//Use HiveDeserializer to deserialize an object out of a Writable blob
 			Object hiveRowStruct = deserializer.deserialize(value);
-			int index = 0;
-			for (; index < structFields.size(); index++) {
-				StructField structField = structFields.get(index);
-				Object object = HiveInspectors.toFlinkObject(structField.getFieldObjectInspector(),
-						structObjectInspector.getStructFieldData(hiveRowStruct, structField));
-				row.setField(index, object);
-			}
-			for (String partition : partitionColNames){
-				row.setField(index++, hiveTablePartition.getPartitionSpec().get(partition));
+			for (int i = 0; i < fields.length; i++) {
+				// non-partition column
+				if (fields[i] < structFields.size()) {
+					StructField structField = structFields.get(fields[i]);
+					Object object = HiveInspectors.toFlinkObject(structField.getFieldObjectInspector(),
+							structObjectInspector.getStructFieldData(hiveRowStruct, structField));
+					row.setField(i, object);
+				} else {
+					// partition column
+					String partition = partitionColNames.get(fields[i] - structFields.size());
```

Comment:
minor: the partition column names seem can be precomputed as they stay the same for rows in the same partition?

Reply:
Yeah, will pre-compute the partition values.